src/THORChain/Swap.cpp

```diff
@@ -32,30 +32,49 @@
 
 namespace TW::THORChainSwap {
 
+static Data ethAddressStringToData(const std::string& asString) {
+    Data asData(20);
+    if (asString.empty() || !Ethereum::Address::isValid(asString)) {
+        return asData;
+    }
+    auto address = Ethereum::Address(asString);
+    std::copy(address.bytes.begin(), address.bytes.end(), asData.data());
+    return asData;
+}
+
 TWCoinType chainCoinType(Chain chain) {
     switch (chain) {
-        case Chain::ETH: return TWCoinTypeEthereum;
-        case Chain::BNB: return TWCoinTypeBinance;
-        case Chain::BTC: return TWCoinTypeBitcoin;
-        case Chain::THOR:
-        default:
-            return TWCoinTypeTHORChain;
+    case Chain::ETH:
+        return TWCoinTypeEthereum;
+    case Chain::BNB:
+        return TWCoinTypeBinance;
+    case Chain::BTC:
+        return TWCoinTypeBitcoin;
+    case Chain::THOR:
+    default:
+        return TWCoinTypeTHORChain;
     }
 }
 
 std::string chainName(Chain chain) {
     switch (chain) {
-        case Chain::ETH: return "ETH";
-        case Chain::BNB: return "BNB";
-        case Chain::BTC: return "BTC";
-        case Chain::THOR:
-        default:
-            return "THOR";
+    case Chain::ETH:
+        return "ETH";
+    case Chain::BNB:
+        return "BNB";
+    case Chain::BTC:
+        return "BTC";
+    case Chain::THOR:
+    default:
+        return "THOR";
     }
 }
 
-std::string Swap::buildMemo(Chain toChain, const std::string& toSymbol, const std::string& toTokenId, const std::string& toAddress, uint64_t limit, const std::string& feeAddress, std::optional<uint16_t> feeRate, const std::string& extra) {
+std::string Swap::buildMemo(Proto::Asset toAsset, const std::string& toAddress, uint64_t limit, const std::string& feeAddress, std::optional<uint16_t> feeRate, const std::string& extra) {
     std::string prefix = "SWAP";
+    const auto& toChain = static_cast<Chain>(toAsset.chain());
+    const auto& toTokenId = toAsset.token_id();
+    const auto& toSymbol = toAsset.symbol();
     if (toChain == Chain::ETH) {
         prefix = "=";
     }
```

Comment:
I think this old `buildMemo` is duplicate, should be removed.

Reply:
Yes I'm thinking of a way to do the UT with the swap builder